lint/initClause_checker.go

```diff
@@ -0,0 +1,51 @@
+package lint
+
+import (
+	"go/ast"
+
+	"github.com/go-toolsmith/astp"
+)
+
+func init() {
+	addChecker(&initClauseChecker{}, attrSyntaxOnly, attrExperimental)
+}
+
+type initClauseChecker struct {
+	checkerBase
+}
+
+func (c *initClauseChecker) InitDocumentation(d *Documentation) {
+	d.Summary = "Detects non-assignment statements inside if/switch init clause."
+	d.Before = `if sideEffect(); cond {
+}`
+	d.After = `sideEffect()
+if cond {
+}`
+}
+
+func (c *initClauseChecker) VisitStmt(stmt ast.Stmt) {
+	if expr, ok := c.getInitClause(stmt).(*ast.ExprStmt); ok {
+		if !astp.IsAssignStmt(expr.X) {
```

Comment:
I'm concerned.
Does this work?

Reply:
Sure, why not?

Reply:
Actually, you can do this:
```go
if !astp.IsAssignStmt(c.getInitClause(stmt)) {
// warn
}
```

Reply:
I'm concerned because `ast.ExprStmt` can't have `ast.AssignStmt` in it's `X` field. Ever.
This is confusing piece of code that looks like unreachable.